File: github-action/src/extract.py
"""Regex-based string extraction for JSX / TSX / HTML.

This is deliberately a v1 extractor — regex, not AST. BUILD_PLAN §15
upgrades to a proper TypeScript AST walk later. For v1 we want to
cover enough to be useful without turning into a TS compiler
shipped in a Docker image.

Two kinds of string land in `extract_strings`:

1. **JSXText** — the text node between tags.
   `<h1>Welcome back</h1>` → `"Welcome back"`

2. **String-literal attributes** — specific attrs where the value is
   almost always UI copy.
   `<img alt="User avatar">` → `"User avatar"`
   `<input placeholder="Enter your email">` → `"Enter your email"`

We skip anything that looks dynamic (template literals, interpolations,
identifiers). We skip strings that are too short to be meaningful
(< 2 characters), all-punctuation, or purely numeric — reduces noise
without missing legitimate copy.

Any false positives here become false-positive violations later. Worse
than too-permissive is too-strict; the downstream check is still the
content checker, which has its own guardrails.
"""

from __future__ import annotations

import fnmatch
import re
from dataclasses import dataclass
from pathlib import Path


# Attributes that almost always hold user-facing copy. If an attribute
# is missing from this list, we skip its string value. Favor precision
# over recall for v1 — false positives are a worse UX than missed strings.
COPY_ATTRIBUTES = frozenset(
    {
        "alt",
        "aria-label",
        "arialabel",  # some codebases drop the hyphen
        "label",
        "placeholder",
        "title",
        "description",
        "tooltip",
        "subtitle",
        "heading",
        "text",
        "message",
        "content",
    }
)

# Tags whose inner text is almost never user-facing — skip their JSXText.
SKIP_INNER_TEXT_TAGS = frozenset({"script", "style", "noscript", "code", "pre"})


_ATTR_RE = re.compile(
    r"""
    \b(?P<attr>[a-zA-Z][a-zA-Z0-9_-]*)   # attribute name
    \s*=\s*
    "(?P<value>[^"]*)"                    # double-quoted value only;
                                          # single-quoted is also valid HTML
                                          # but is handled separately to
                                          # keep the regex readable
    """,
    re.VERBOSE,
)

_ATTR_RE_SINGLE = re.compile(
    r"\b(?P<attr>[a-zA-Z][a-zA-Z0-9_-]*)\s*=\s*'(?P<value>[^']*)'"
)

# Very loose JSX element tokenizer: grab the tag open, the attrs span,
# and the text immediately after. Multi-line-friendly. Not a parser.
_JSX_TEXT_RE = re.compile(
    r"<(?P<tag>[A-Za-z][A-Za-z0-9_.-]*)\b[^>]*>(?P<text>[^<]+?)</",
    re.DOTALL,
)


@dataclass(frozen=True)
class Extraction:
    """One string pulled from source. `kind` distinguishes how it was found."""

    text: str
    source_file: str
    line: int
    kind: str  # "jsx-text" or "attr:<name>"


def extract_strings(path: Path, source: str | None = None) -> list[Extraction]:
    """Pull all UI-copy candidates from a single source file."""
    source = source if source is not None else path.read_text(encoding="utf-8")
    hits: list[Extraction] = []

    # ------------------------------------------------------------------
    # Pass 1: JSX / HTML text nodes.
    # ------------------------------------------------------------------
    for match in _JSX_TEXT_RE.finditer(source):
        tag = match.group("tag").lower()
        if tag in SKIP_INNER_TEXT_TAGS:
            continue
        raw_text = match.group("text")
        for piece in _split_and_clean(raw_text):
            if not _looks_like_copy(piece):
                continue
            line = source.count("\n", 0, match.start("text")) + 1
            hits.append(
                Extraction(text=piece, source_file=str(path), line=line, kind="jsx-text")
            )

    # ------------------------------------------------------------------
    # Pass 2: attribute values. Both quote styles.
    # ------------------------------------------------------------------
    for regex in (_ATTR_RE, _ATTR_RE_SINGLE):
        for match in regex.finditer(source):
            attr = match.group("attr").lower()
            if attr not in COPY_ATTRIBUTES:
                continue
            value = match.group("value")
            if not _looks_like_copy(value):
                continue
            if _looks_dynamic(value):
                continue
            line = source.count("\n", 0, match.start("value")) + 1
            hits.append(
                Extraction(
                    text=value,
                    source_file=str(path),
                    line=line,
                    kind=f"attr:{attr}",
                )
            )

    return _dedupe(hits)


def matches_glob(path: str, pattern: str) -> bool:
    """Glob match with brace expansion (**/*.{tsx,jsx,html}).

    Python's fnmatch doesn't do `{a,b,c}`, so we expand manually and
    OR the results together.
    """
    for expanded in _expand_braces(pattern):
        if fnmatch.fnmatch(path, expanded):
            return True
    return False


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------
def _split_and_clean(raw: str) -> list[str]:
    """Collapse whitespace; return a single trimmed string per JSXText node."""
    cleaned = re.sub(r"\s+", " ", raw).strip()
    if not cleaned:
        return []
    return [cleaned]


def _looks_like_copy(text: str) -> bool:
    """Heuristic: is this string likely user-visible UI copy?

    Reject:
      - too short (< 2 chars)
      - all digits / all punctuation
      - starts with `{` or `$` (interpolation leftover)
      - contains backtick (template literal)
      - a single word that looks like an identifier (camelCase/snake_case)
    """
    if len(text) < 2:
        return False
    if text.strip().startswith("{"):
        return False
    if "`" in text or "${" in text:
        return False
    if not re.search(r"[A-Za-z]", text):
        return False
    if " " not in text and re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", text):
        # single identifier-looking token — likely a variable reference
        return False
    if re.fullmatch(r"[\s\d\W_]+", text):
        return False
    return True


def _looks_dynamic(value: str) -> bool:
    # Catches things like title="Welcome, ${user.name}" that squeaked
    # past the attribute regex because the $ is inside the quotes.
    return "${" in value or "{{" in value


def _dedupe(hits: list[Extraction]) -> list[Extraction]:
    """Keep the first occurrence of each (file, line, text) — regexes
    sometimes double-match near attribute/text boundaries."""
    seen = set()
    out = []
    for h in hits:
        key = (h.source_file, h.line, h.text, h.kind)
        if key in seen:
            continue
        seen.add(key)
        out.append(h)
    return out


def _expand_braces(pattern: str) -> list[str]:
    """`a{b,c}d` → `['abd', 'acd']`. Only handles one level of braces,
    which is all BUILD_PLAN §12's default pattern needs."""
    match = re.search(r"\{([^{}]+)\}", pattern)
    if not match:
        return [pattern]
    alternatives = match.group(1).split(",")
    prefix = pattern[: match.start()]
    suffix = pattern[match.end() :]
    return [prefix + alt + suffix for alt in alternatives]

File: github-action/src/extract.mjs
#!/usr/bin/env node
/**
 * extract.mjs — Babel-based UI-copy extractor.
 *
 * Reads one or more file paths from argv (absolute or repo-relative),
 * parses each with @babel/parser (JSX/TSX plugin) and walks the AST
 * with @babel/traverse, emitting one JSON object per extracted string
 * to stdout (NDJSON — newline-delimited JSON). The Python side reads
 * stdin line-by-line and issues `contentrx` calls.
 *
 * This replaces the regex extractor in src/extract.py for .jsx/.tsx.
 * HTML files are still handled by the regex path inside main.py so we
 * don't need htmlparser2 in the image today. Migrate HTML to AST in a
 * follow-up if the regex coverage there stops being enough.
 *
 * Output schema (one per line):
 *   {
 *     "text":    string,
 *     "file":    string,
 *     "line":    integer,  // 1-indexed
 *     "column":  integer,  // 0-indexed, Babel convention
 *     "kind":    string,   // "jsx-text" | "attr:<name>" | "template-literal"
 *   }
 *
 * Extraction rules (locked per BUILD_PLAN §15):
 *   - JSXText with trimmed length > 3, not whitespace-only
 *   - String literals in JSX attributes whose name matches COPY_ATTRS
 *   - Template literal STATIC parts only — skip any ${…} interpolations
 *   - Skip strings that look like URLs, paths, CSS classes, identifiers,
 *     or pure punctuation / digits
 */

import fs from "node:fs";
import path from "node:path";
import { parse } from "@babel/parser";
import _traverse from "@babel/traverse";

// @babel/traverse is CJS-shipped; the default export is the callable
// function, but under ESM the module namespace wraps it.
const traverse = _traverse.default ?? _traverse;

const COPY_ATTRS = new Set([
  "alt",
  "ariaLabel",
  "aria-label",
  "placeholder",
  "title",
  "label",
  "description",
  "helperText",
  "errorMessage",
  "children",
  "heading",
  "subtitle",
  "tooltip",
  "message",
  "content",
]);

const MIN_LENGTH = 3;

// Patterns we reject outright — they're almost never user-facing copy.
const URL_RE = /^https?:\/\//i;
const PATH_RE = /^(?:\.\.?\/|\/|~\/)/;
const CSS_CLASS_RE = /^[a-z][\w-]*(?:\s+[a-z][\w-]*)+$/;     // "btn btn-primary"
const SINGLE_TOKEN_RE = /^[A-Za-z_][A-Za-z0-9_-]*$/;          // "userName", "btn-primary"
const PUNCT_ONLY_RE = /^[\s\d\W_]+$/;

function looksLikeCopy(raw) {
  const text = raw.trim();
  if (text.length < MIN_LENGTH) return false;
  if (URL_RE.test(text)) return false;
  if (PATH_RE.test(text)) return false;
  if (PUNCT_ONLY_RE.test(text)) return false;
  if (!/[A-Za-z]/.test(text)) return false;
  // CSS-class-ish ("btn btn-primary") — no spaces inside single-token-like words.
  if (!/[\s]/.test(text) && SINGLE_TOKEN_RE.test(text)) return false;
  if (/^\s*$/.test(text)) return false;
  if (CSS_CLASS_RE.test(text) && !/[.?!,;:]/.test(text)) return false;
  return true;
}

function normaliseAttrName(name) {
  if (typeof name === "string") return name.toLowerCase();
  if (name && typeof name === "object" && name.type === "JSXNamespacedName") {
    return `${name.namespace.name}:${name.name.name}`.toLowerCase();
  }
  if (name && typeof name === "object" && name.name) return String(name.name).toLowerCase();
  return "";
}

function extractFromSource(filePath, source) {
  const hits = [];
  let ast;
  try {
    ast = parse(source, {
      sourceType: "unambiguous",
      plugins: [
        "jsx",
        "typescript",
        "classProperties",
        "decorators-legacy",
        "topLevelAwait",
        "importAssertions",
      ],
      errorRecovery: true,
    });
  } catch (err) {
    process.stderr.write(
      `warning: failed to parse ${filePath}: ${err.message}\n`,
    );
    return hits;
  }

  function push(text, loc, kind) {
    const trimmed = text.trim().replace(/\s+/g, " ");
    if (!looksLikeCopy(trimmed)) return;
    hits.push({
      text: trimmed,
      file: filePath,
      line: loc?.start?.line ?? 1,
      column: loc?.start?.column ?? 0,
      kind,
    });
  }

  traverse(ast, {
    JSXText(p) {
      push(p.node.value, p.node.loc, "jsx-text");
    },
    JSXAttribute(p) {
      const attrName = normaliseAttrName(p.node.name?.name ?? p.node.name);
      if (!COPY_ATTRS.has(attrName)) return;
      const value = p.node.value;
      if (!value) return;
      if (value.type === "StringLiteral") {
        push(value.value, value.loc, `attr:${attrName}`);
      } else if (value.type === "JSXExpressionContainer") {
        const expr = value.expression;
        if (expr.type === "StringLiteral") {
          push(expr.value, expr.loc, `attr:${attrName}`);
        } else if (expr.type === "TemplateLiteral") {
          // Emit every static quasi (e.g. `Hello ${name}, welcome!` →
          // "Hello" and "welcome!"). Interpolated slots are skipped.
          for (const quasi of expr.quasis) {
            push(quasi.value.cooked ?? quasi.value.raw ?? "", quasi.loc, "template-literal");
          }
        }
      }
    },
    // Standalone TemplateLiteral outside of a JSX attribute — common
    // in things like toast("Welcome back"). We only capture when the
    // template has no interpolations; otherwise we'd hand the AI copy
    // with placeholder gaps and over-flag.
    TemplateLiteral(p) {
      if (p.node.expressions.length > 0) return;
      if (p.parent.type === "JSXAttribute") return;
      if (p.parent.type === "JSXExpressionContainer") return; // handled above
      const quasi = p.node.quasis[0];
      if (!quasi) return;
      push(
        quasi.value.cooked ?? quasi.value.raw ?? "",
        p.node.loc,
        "template-literal",
      );
    },
  });

  return hits;
}

function extractFromPath(absPath, repoRoot) {
  const repoRelative = path.relative(repoRoot, absPath);
  let source;
  try {
    source = fs.readFileSync(absPath, "utf-8");
  } catch (err) {
    process.stderr.write(`warning: could not read ${absPath}: ${err.message}\n`);
    return [];
  }
  return extractFromSource(repoRelative || absPath, source);
}

function main() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    process.stderr.write("usage: extract.mjs <file> [file …]\n");
    process.exit(2);
  }

  const repoRoot = process.env.GITHUB_WORKSPACE || process.cwd();

  for (const arg of args) {
    const absPath = path.isAbsolute(arg) ? arg : path.resolve(repoRoot, arg);
    const hits = extractFromPath(absPath, repoRoot);
    for (const h of hits) {
      process.stdout.write(`${JSON.stringify(h)}\n`);
    }
  }
}

main();
